File: mines.js
var Board = function(r, c, totalMines) {
  this.rows = r;
  this.cols = c;
  this.cells = [];
  this.totalMines = totalMines;
  reset();
};

Board.STATUS_ENDED    = 1;
Board.STATUS_STARTED  = 0;

Board.prototype.reset = function() {
  var gridSize = this.rows * this.cols, i;
  for(i = 0; i < gridSize; i++) {
    this.cells.puth(new Cell());
  }  
  this.placeMines();
}

Board.prototype.turnOnCell = function(index) {
  var cell = this.cells[index];

  if(cell.status == Cell.STATUS_ON)    
    return;
  
  cell.turnOn();  
    
  if(cell.type == Cell.TYPE_MINE) {
    this.status = Board.STATUS_ENDED;
    return;
  }
  
  if(cell.type == Cell.TYPE_EMPTY) {
    this.turnOnAdjacentCells(index);
  }    
}

Board.prototype.turnOnAdjacentCells = function(index) {
  var cellIndexes = this.getAdjacentCellIndexes(index),
      i = 0,
      cell = null;

  for(i = 0; i < cellIndexes.length; i++) {
    cell = this.cells[i];

    if(cell.status == Cell.STATUS_ON) {
      continue;
    }

    if(cell.type == Cell.TYPE_EMPTY) {
      cell.turnOn();
      this.turnOnAdjacentCells(i);      
    } else if(cell.type == Cell.TYPE_MARKED) {
      cell.turnOn();
    }
  }
}

Board.prototype.placeMines = function() {  
    // Creates an array with each element initialized to its index value.
    var gridIndexes = Array.apply(null, {length: cells.length}).map(Number.call, Number),
        minePositions = sample(gridIndexes, this.totalMines),
        i = 0;
    
    for(i = 0; i < minePositions.length; i++) {
      cells[i].type = Cell.TYPE_MINE;
      this.markAdjacentCells(i);
    }    
}

Board.prototype.markAdjacentCells = function(index) {
  var adjacentCellIndexes = this.getAdjacentCellIndexes(index),
      i = 0;

  for(i = 0; i < adjacentCellIndexes.length; i++) {
    if(this.canMarkCell(i)) {
      this.cells[i].mark();
    }    
  }
}

Board.prototype.getAdjacentCellIndexes = function(index) {
    var tmp = this.getRowCol(index),
        row = tmp.row,
        col = tmp.col;

    cellIndexes = [];
    cellIndexes.push(this.getRelativeIndex(row, col, 'above'));
    cellIndexes.push(this.getRelativeIndex(row, col, 'bottom'));    
    cellIndexes.push(this.getRelativeIndex(row, col, 'left'));    
    cellIndexes.push(this.getRelativeIndex(row, col, 'right'));
    cellIndexes.push(this.getRelativeIndex(row, col, 'above_left'));
    cellIndexes.push(this.getRelativeIndex(row, col, 'above_right'));
    cellIndexes.push(this.getRelativeIndex(row, col, 'bottom_left'));
    cellIndexes.push(this.getRelativeIndex(row, col, 'bottom_right'));
    return cellIndexes;
}

Board.prototype.getRowCol = function(fromIndex) {
  var r = fromIndex /this.cols,
      c = fromIndex % this.cols;

  return {
    row: r + 1, 
    col: c + 1
  };
}

Board.prototype.getCell = function(r, c) {  
    return (r -1) * this.cols + (c -1);
}

Board.prototype.getRelativeIndex = function(row, col, direction) {
  var relativeIndex = -1;
  switch(direction) {
    case 'above':
      relativeIndex = row - 1 > 0 ? this.getCell(row - 1, col) : -1;
      break;
    case 'bottom':
      relativeIndex = row + 1 <= this.rows ? this.getCell(row + 1, col) : -1;
      break;
    case 'left':
      relativeIndex = col - 1 > 0 ? this.getCell(row, col - 1)  : -1;
      break;
    case 'right':
      relativeIndex = col + 1 <= this.cols ? this.getCell(row, col + 1)  : -1;
      break;
    case 'above_left':
      relativeIndex = (row - 1 > 0 && col - 1 > 0) ? this.getCell(row - 1, col - 1)  : -1;
      break;
    case 'above_right':
      relativeIndex = (row - 1 > 0 && col + 1 <= this.cols) ? this.getCell(row - 1, col + 1)  : -1;
      break;
    case 'bottom_left':
      relativeIndex = (row + 1 <= this.rows && col - 1 > 0) ? this.getCell(row + 1, col - 1)  : -1;
      break;
    case 'bottom_right':
      relativeIndex = (row + 1 <= this.rows && col +1 <= this.cols) ? this.getCell(row + 1, col + 1)   : -1
      break;
  }

  return relativeIndex;
}

Board.prototype.canMarkCell = function(index) {
    return !isOutOfBounds(index) && this.cells[index].type != Cell.TYPE_MINE;
}

Board.prototype.isOutOfBounds = function(index) {
  return index < 0 || index >= this.cells.length;
}


// Cell class
var Cell = function() {
  this.type = type;
  this.tag = 0;
  this.status = Cell.STATUS_OFF
}

Cell.STATUS_OFF   = 0;
Cell.STATUS_ON    = 1;
Cell.TYPE_EMPTY   = 0;
Cell.TYPE_MINE    = 1;
Cell.TYPE_MARKED  = 2;

Cell.prototype.mark = function() {
  this.type = Cell.TYPE_MARKED;
  this.tag = this.tag + 1;
}

Cell.prototype.turnOn = function() {
  this.status = Cell.STATUS_ON
}

Cell.prototype.display = function() {
  return ' ' if this.status == Cell.STATUS_OFF;
  switch(this.type) {
    case Cell.TYPE_MINE:
      return '* ';
      break;
    case Cell.TYPE_MARKED
      return this.tag.toString();
      break;
    default:
      return "0";
  }
}


function sample(array, size) {
    var tmp, current, top = array.length;

    if(top) while(--top) {
      current = Math.floor(Math.random() * (top + 1));
      tmp = array[current];
      array[current] = array[top];
      array[top] = tmp;
    }

    return array.slice(0, size);
}

File: mines.rb
require 'pp'

class Board
	STATUS_STARTED = 0
	STATUS_ENDED = 1
	
	attr_accessor :cells, :total_mines, :rows, :cols
	attr_reader :status
	def initialize(r, c, total_mines)
		self.rows = r
		self.cols = c
		self.cells = []
		self.total_mines = total_mines
		reset
	end

	def reset		
		(rows * cols).times do |index|
			cells << Cell.new
		end
		place_mines
	end
	
	def turn_on_cell(index)
		cell = cells[index]
		return if cell.status == Cell::STATUS_ON		
		
		cell.turn_on
		if cell.type == Cell::TYPE_MINE
			@status = STATUS_ENDED
			return
		end
		
		if cell.type == Cell::TYPE_EMPTY
			turn_on_adjacent_cells(index)
		end
	end
	
	def turn_on_adjacent_cells(from_index)
		cell_indexes = get_adjacent_cell_indexes(from_index)
		cell_indexes.each do |cell_index|
			cell = cells[cell_index]
			next if cell.status == Cell::STATUS_ON
			if cell.type == Cell::TYPE_EMPTY
				cell.turn_on
				turn_on_adjacent_cells(cell_index)
			elsif cell.type == Cell::TYPE_MARKED
				cell.turn_on
			end
		end
	end
	
	def hint
	end
	
	def get_adjacent_cell_indexes(from_index)
		row, col = get_row_col(from_index)
		cell_indexes = []
		cell_indexes << get_relative_index(row,col, :above)
		cell_indexes << get_relative_index(row,col, :bottom)
		cell_indexes << get_relative_index(row,col, :left)
		cell_indexes << get_relative_index(row,col, :right)
		cell_indexes << get_relative_index(row,col, :above_left)
		cell_indexes << get_relative_index(row,col, :above_right)
		cell_indexes << get_relative_index(row,col, :bottom_left)
		cell_indexes << get_relative_index(row,col, :bottom_right)
		cell_indexes
	end
	
	def place_mines
		#~ mine_positions = (0..cells.length - 1).to_a.sample total_mines				
		mine_positions = (1..cells.length - 1).to_a.sample total_mines
		mine_positions = [55, 53, 6, 17, 57, 58, 34, 40, 25, 41]
		p mine_positions
		mine_positions.each do |index|
			cells[index].type = Cell::TYPE_MINE			
			mark_adjacent_cells(index)
		end		
	end
	
	def mark_adjacent_cells(index)		
		get_adjacent_cell_indexes(index).each do |cell_index|
			cells[cell_index].mark if can_mark_cell?(cell_index)
		end

		#~ p ""
	end
	
	def display
		rows.times do |r|
			v = '|'
			cols.times do |c|
				v += "     " + cells[get_cell(r + 1,c + 1)].display + "    |"				
			end
			p v
		end
	end
	
	private
	
	def get_relative_index(row, col, direction)
		case direction 
		when :above
			row - 1 > 0 ? get_cell(row - 1, col) : -1
		when :bottom
			row + 1 <= rows ? get_cell(row + 1, col) : -1
		when :left
			col - 1 > 0 ? get_cell(row, col - 1)  : -1
		when :right
			col + 1 <= cols ? get_cell(row, col + 1)  : -1
		when :above_left
			(row - 1 > 0 && col - 1 > 0) ? get_cell(row - 1, col - 1)  : -1
		when :above_right
			(row - 1 > 0 && col + 1 <= cols) ? get_cell(row - 1, col + 1)  : -1
		when :bottom_left
			(row + 1 <= rows && col - 1 > 0) ? get_cell(row + 1, col - 1)  : -1
		when :bottom_right
			(row + 1 <= rows && col +1 <= cols) ? get_cell(row + 1, col + 1)   : -1						 
		end
	end
	
	def can_mark_cell?(index)
		!out_of_bounds?(index) && cells[index].type != Cell::TYPE_MINE
	end
	
	def out_of_bounds?(index)		
		index < 0 || index >= cells.length
	end
	
	def get_row_col(index)
		r = index/cols
		c = index % cols
		[r + 1, c + 1]
	end
	
	def get_cell(r,c)
		(r -1) * cols + (c -1)
	end
end

class Cell	
	TYPE_EMPTY 	= 0
	TYPE_MINE	= 1
	TYPE_MARKED	= 2
	
	STATUS_OFF = 0
	STATUS_ON = 1
	
	attr_accessor :type, :tag, :status
	def initialize(type = TYPE_EMPTY)
		self.type = type
		self.tag = 0
		self.status = STATUS_OFF
	end
	
	def mark
		self.type = TYPE_MARKED
		self.tag += 1
	end
	
	def turn_on
		self.status = STATUS_ON
	end
	
	def display
		return ' ' if status == STATUS_OFF
		case type
		when TYPE_MINE
			'* '
		when TYPE_MARKED
			self.tag.to_s
		else
			'0'
		end
	end
end

ROWS = 8
COLS = 8
MINES = 10
board = Board.new(ROWS, COLS, MINES)
board.display
p ""
board.turn_on_cell(0)
board.display
p ""
board.turn_on_cell(16)
board.display

p ""
board.turn_on_cell(17)
board.display
p board.status